README.md
<!--- coding: utf-8; fill-column: 80 --->
# ![ligō](https://github.com/nibua-r/ligo-logos/raw/master/logo/ligo.png)

[![Dependency Status](https://gemnasium.com/nibua-r/ligo.png)](https://gemnasium.com/nibua-r/ligo)
[![Code Climate](https://codeclimate.com/github/nibua-r/ligo.png)](https://codeclimate.com/github/nibua-r/ligo)

* [Homepage](https://github.com/nibua-r/ligo#readme)
* [Issues](https://github.com/nibua-r/ligo/issues)
* [Documentation](http://rubydoc.info/gems/ligo/frames)
* [Email](mailto:root at renaud.io)

## Description

ligō is a ruby gem implementing the
[Android Open Accessory Protocol](http://source.android.com/tech/accessories/aoap/aoa.html)
to interact with Android devices via USB.

Android-side application is needed to make some use of this project. A
[sample LigoTextDemo Android application](https://github.com/nibua-r/LigoTextDemo)
is available for usage demonstration.

## Features

Just USB I/O using the Android Open Accessory Protocol…

## Examples

```ruby
require 'ligo'

id = {
  manufacturer: 'Ligo',
  model:        'VirtualAccessory',
  description:  'PC acts as an accessory!',
  version:      '0.1',
  uri:          'http://blog.renaud.io',
  serial:       '3678000012345678'
}

Ligo::Logging.configure_logger_output('/tmp/ligo-sample.log')
ctx = Ligo::Context.new
acc = Ligo::Accessory.new(id)

# You could inspect the Android devices list
puts ctx.devices.inspect

# Take the first one
dev = ctx.devices.first

# Switch the Android device to accessory mode
success = dev.attach_accessory(acc)

# then operate on the device (you'll need some Android code for this!)
if success
  dev.process do |handle|
    # I/O
  end
end
```

## Requirements

ligō will be nothing without the
[LIBUSB Ruby binding](https://github.com/larskanis/libusb) and the underlying
[libusb](http://libusbx.org/) library.

At the time of writing, ligō focuses on the GNU/Linux operating system and
YARV/MRI ruby. ligō has been developed and tested on Debian GNU/Linux with ruby
1.9.3p327 and 1.9.3p429.

ligō is useless without Android-side application. A
[sample LigoTextDemo Android application](https://github.com/nibua-r/LigoTextDemo)
is available for usage demonstration.

## Install

You may install ligō by using:

    $ gem install ligo

Or, you can install it from the github repository:

    $ git clone https://github.com/nibua-r/ligo.git
    $ cd ligo
    $ rake install


## Pending tasks

* Write the specs and automate testing!
* ~~Provide an Android demo application~~ → See
  [sample LigoTextDemo Android application](https://github.com/nibua-r/LigoTextDemo),
  more to come soon…
* Improve the documentation → in progress, a first pass has been made…
* ~~Release and push the gem to rubygems.org~~
* Finalize and publish a [celluloid](https://github.com/celluloid/celluloid)-based write-back usage
  sample (already working with the
  [sample LigoTextDemo Android application](https://github.com/nibua-r/LigoTextDemo) and the
  [libusb 0.3.0 gem](https://github.com/larskanis/libusb) async and polling APIs)
* Create a wiki with FAQ, supported devices listing, UMS/accessory interactions, etc.

## Contributing to ligō

* Follow the usual fork/branch/PR workflow to send changes, if I like them I'll merge them
* Help me to create a supported device listing

## Copyright

Copyright (c) 2012, 2013 Renaud AUBIN

See {file:LICENSE.txt} for details.


ChangeLog.md
### 0.2.0 / 2016-03-18

* Enable to configure the reenumeration delay (i.e. when the device switches from nominal to accessory mode by re-introduce itself on the USB bus)

### 0.1.2 / 2013-11-14

* Add Windows support
* Improve the accessory_mode? method for devices which expose the Google vendorId even when not in accessory mode

### 0.1.1 / 2013-04-13

* Fix a major issue due to bad refactoring (get_device should return to device)
* Upgrade dependencies

### 0.1.0 / 2012-12-24

* Initial release


spec/ligo_spec.rb
require 'spec_helper'

describe Ligo do
  it "should have a VERSION constant" do
    subject.const_get('VERSION').should_not be_empty
  end
end


spec/spec_helper.rb
gem 'rspec', '~> 2.4'
require 'rspec'
require 'ligo'

include Ligo
Ligo::Logging.configure_logger_output('/tmp/ligo-spec.log')


spec/ligo/device_spec.rb
require 'spec_helper'

# Assumption: two devices are connected on the usb bus and not yet in accessory
# mode. Those devices are a Galaxy Nexus running 4.2 AOSP and a HTC Flyer
# running 3.2 (with UMS support).

describe Ligo::Device do
  before(:all) do
    @default_accessory = Ligo::Accessory.new
    ctx = Ligo::Context.new
    @gnexus = ctx.devices(idVendor: 0x04e8).first
    @flyer  = ctx.devices(idVendor: 0x0bb4).first
  end

  it 'should derive from LIBUSB::Device' do
    Ligo::Device.superclass.should be LIBUSB::Device
  end

  context 'when passing the Galaxy Nexus to accessory mode' do
    specify { @gnexus.should_not be_accessory_mode }
    specify { @gnexus.idVendor.should  be 0x04e8 }
    specify { @gnexus.idProduct.should be 0x6860 }
    specify { @gnexus.should be_aoap }
    specify { @gnexus.should_not be_uas }
    specify { @gnexus.aoap_version.should be 2 }
    # Now, the order matters!
    specify { @gnexus.attach_accessory(@default_accessory).should be true }
    specify { @gnexus.idVendor.should be 0x18d1 }
    specify { Ligo::GOOGLE_PIDS.should include @gnexus.idProduct }
  end

  context 'when passing the Flyer to accessory mode' do
    specify { @flyer.should_not be_accessory_mode }
    specify { @flyer.idVendor.should  be 0x0bb4 }
    specify { @flyer.idProduct.should be 0x0ca9 }
    specify { @flyer.should be_aoap }
    specify { @flyer.should be_uas }
    specify { @flyer.aoap_version.should be 1 }
    # Now, the order matters!
    specify { @flyer.attach_accessory(@default_accessory).should be true }
    specify { @flyer.idVendor.should be 0x18d1 }
    specify { Ligo::GOOGLE_PIDS.should include @flyer.idProduct }
  end

end


lib/ligo.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'logger'
require 'libusb'

require 'ligo/version'

# This module contains the Android Open Accessory Protocol utility classes to
# enable custom USB I/O with AOAP-compatible devices.
#
# @see http://source.android.com/tech/accessories/index.html
# @see http://source.android.com/tech/accessories/aoap/aoa.html
# @see
#  http://developer.android.com/guide/topics/connectivity/usb/accessory.html
# @author Renaud AUBIN
module Ligo
  require 'ligo/logging'
  require 'ligo/constants'
  require 'ligo/accessory'
  require 'ligo/context'
  require 'ligo/device'

  # Return the OS identifier
  # @return [Symbol] one of :windows, :macosx, :linux or :unix.
  # @raise [StandardError] if the OS is unknown.
  def self.os
    @os ||= (
             host_os = RbConfig::CONFIG['host_os']
             case host_os
             when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
               :windows
             when /darwin|mac os/
               :macosx
             when /linux/
               :linux
             when /solaris|bsd/
               :unix
             else
               raise StandardError, "unknown os: #{host_os.inspect}"
             end
             )
  end
end


lib/ligo/device.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Ligo

  require 'ligo/constants'

  # USB reenumeration delay in seconds
  @@reenumeration_delay = 1

  def self.setDelay(x)
    @@reenumeration_delay = x
  end

  def self.getDelay
    @@reenumeration_delay
  end

  # This class provides a convenient wrapper class around `LIBUSB::Device` and
  #   implements the Android Open Accessory Protocol to interact with compatible
  #   devices.
  #
  # This class is a derivative work of `LIBUSB::Device` as included in
  #   [LIBUSB](https://github.com/larskanis/libusb), written by Lars Kanis and
  #   released under the LGPLv3.
  # @author Renaud AUBIN
  # @api public
  class Device < LIBUSB::Device
    include Logging

    # @api private
    attr_reader :pDev

    # @api private
    attr_reader :pDevDesc

    # Returns the version of the AOA protocol that this device supports
    # @return [Fixnum] the version of the AOA protocol that this device
    #   supports.
    attr_reader :aoap_version

    # Returns the associated {Accessory}
    # @return [Accessory, nil] the associated accessory if any or nil.
    attr_reader :accessory

    # Returns the accessory mode input endpoint
    # @return [LIBUSB::Endpoint, nil] the input endpoint or nil if the device is
    #   not in accessory mode.
    attr_reader :in

    # Returns the accessory mode output endpoint
    # @return [LIBUSB::Endpoint, nil] the output endpoint or nil if the device
    #   is not in accessory mode.
    attr_reader :out

    # Returns the device handle
    # @todo Improve the :handle doc
    # @return [LIBUSB::DevHandle, nil] the device handle or nil.
    attr_reader :handle

    # @api private
    def initialize context, pDev
      @aoap_version = 0
      @accessory, @in, @out, @handle = nil, nil, nil, nil
      super context, pDev
    end

    def process(&block)
      begin
        self.open_interface(0) do |handle|
          @handle = handle
          yield handle
          @handle = nil
        end
        # close
      rescue LIBUSB::ERROR_NO_DEVICE
        msg =  'The target device has been disconnected'
        logger.debug msg
        # close
        raise Interrupt, msg
      end
    end

    # Opens an handle and claim the default interface for further operations
    # @return [LIBUSB::DevHandle] the handle to operate on.
    # @raise
    def open_and_claim
      @handle = open
      @handle.claim_interface(0)
      @handle.clear_halt(@in)
      @handle
    end

    # Finalizes the device (release and close)
    # @return
    # @raise [LIBUSB::ERROR_TIMEOUT] in case of timeout.
    def finalize
      if @handle
        @handle.release_interface(0)
        @handle.close
      end
    end

    # Simple write method (blocking until timeout)
    # @param [Fixnum] buffer_size
    #   The number of bytes expected to be received.
    # @param [Fixnum] timeout
    #   The timeout in ms (default: 1000). 0 for an infinite timeout.
    # @return [String] the received buffer (at most buffer_size bytes).
    # @raise [LIBUSB::ERROR_TIMEOUT] in case of timeout.
    def read(buffer_size, timeout = 1000)
      handle.bulk_transfer(endpoint: @in,
                           dataIn: buffer_size,
                           timeout: timeout)
    end
    alias_method :recv, :read

    # Simple write method (blocking until timeout)
    # @param [String] buffer
    #   The buffer to be sent.
    # @param [Fixnum] timeout
    #   The timeout in ms (default: 1000). 0 for an infinite timeout.
    # @return [Fixnum] the number of bytes actually sent.
    # @raise [LIBUSB::ERROR_TIMEOUT] in case of timeout.
    def write(buffer, timeout = 1000)
        handle.bulk_transfer(endpoint: @out,
                             dataOut: buffer,
                             timeout: timeout)
    end
    alias_method :send, :write

    # Associates with an accessory and switch to accessory mode
    #
    # Prepare an OAP compatible device to interact with a given {Ligo::Accessory}:
    # * Switch the current assigned device to accessory mode
    # * Set the I/O endpoints
    # @param [Ligo::Accessory] accessory
    #   The virtual accessory to be associated with the Android device.
    # @return [true, false] true for success, false otherwise.
    def attach_accessory(accessory)
      logger.debug "attach_accessory(#{accessory})"

      @accessory = accessory

      if accessory_mode?
        # if the device is already in accessory mode, we send
        # set_configuration to force an usb attached event on the device
        begin
          set_configuration
        rescue LIBUSB::ERROR_NO_DEVICE
          logger.debug '  set_configuration raises LIBUSB::ERROR_NO_DEVICE - Retry'
          sleep Ligo::getDelay
          # Set configuration may fail
          retry
        end
      else
        # the device is not in accessory mode, start_accessory_mode is
        # sufficient to get an usb attached event on the device
        return false unless start_accessory_mode
      end

      # Find out the in/out endpoints
      self.interfaces.first.endpoints.each do |ep|
        if ep.bEndpointAddress & 0b10000000 == 0
          @out = ep if @out.nil?
        else
          @in = ep if @in.nil?
        end
      end
      true
    end

    # Switches to accessory mode
    #
    # Send identifying string information to the device and request the device start up in accessory
    # mode.
    # @return [true, false] true for success, false otherwise.
    def start_accessory_mode
      logger.debug 'start_accessory_mode'
      sn = self.serial_number

      self.open do |handle|
        @handle = handle
        send_accessory_id
        send_start
        @handle = nil
      end

      wait_and_retrieve_by_serial(sn)
    end

    # Sends a `set configuration` control transfer
    #
    # Set the device's configuration to a value of 1 with a SET_CONFIGURATION (0x09) device
    # request.
    # @return [true, false] true for success, false otherwise.
    def set_configuration
      logger.debug 'set_configuration'
      res = nil
      sn = self.serial_number
      device = get_device(sn)

      begin
        device.open_interface(0) do |handle|
          req_type = LIBUSB::ENDPOINT_OUT | LIBUSB::REQUEST_TYPE_STANDARD
          res = handle.control_transfer(bmRequestType: req_type,
                                        bRequest: LIBUSB::REQUEST_SET_CONFIGURATION,
                                        wValue: 1, wIndex: 0x0, dataOut: nil)
        end

        wait_and_retrieve_by_serial(sn)
        res == 0
      end
    end

    # Check if the current {Device} is in accessory mode
    # @return [true, false] true if the {Device} is in accessory mode, false
    #   otherwise.
    def accessory_mode?
      (self.idVendor == GOOGLE_VID) && (GOOGLE_PIDS.include? self.idProduct)
    end

    # Check if the current {Device} supports AOAP
    # @return [true, false] true if the {Ligo::Device} supports AOAP, false
    #   otherwise.
    def aoap?
      @aoap_version = self.get_protocol
      aoap_supported = (@aoap_version >= 1)
      if aoap_supported
        logger.info "#{self.inspect} supports AOA Protocol version #{@aoap_version}."
      else
        logger.info "#{self.inspect} doesn't support AOA Protocol."
      end
      aoap_supported
    end

    # Check if the current {Device} is in UMS mode
    # @return [true, false] true if the {Device} is in UMS mode, false otherwise
    def uas?
      if RUBY_PLATFORM=~/linux/i
        # http://cateee.net/lkddb/web-lkddb/USB_UAS.html
        (self.settings[0].bInterfaceClass == 0x08) &&
          (self.settings[0].bInterfaceSubClass == 0x06)
      else
        false
      end
    end

    # Sends a `get protocol` control transfer
    #
    # Send a 51 control request ("Get Protocol") to figure out if the device
    #   supports the Android accessory protocol. We assume here that the device
    #   has not been opened.
    # @return [Fixnum] the AOAP protocol version supported by the device (0 for
    #   no AOAP support).
    def get_protocol
      logger.debug 'get_protocol'
      res, version = 0, 0
      self.open do |h|

        h.detach_kernel_driver(0) if self.uas? && h.kernel_driver_active?(0)
        req_type = LIBUSB::ENDPOINT_IN | LIBUSB::REQUEST_TYPE_VENDOR
        res = h.control_transfer(bmRequestType: req_type,
                                 bRequest: COMMAND_GETPROTOCOL,
                                 wValue: 0x0, wIndex: 0x0, dataIn: 2)

        version = res.unpack('S')[0]
      end

      (res.size == 2 && version >= 1 ) ? version : 0
    rescue LIBUSB::ERROR_NOT_SUPPORTED, LIBUSB::ERROR_PIPE
      0
    end

    # Sends identifying string information to the device
    #
    # We assume here that the device has already been opened.
    # @api private
    # @return
    def send_accessory_id
      logger.debug 'send_accessory_id'
      req_type = LIBUSB::ENDPOINT_OUT | LIBUSB::REQUEST_TYPE_VENDOR
      @accessory.each do |k,v|
        # Ensure the string is terminated by a null char
        s = "#{v}\0"
        r = @handle.control_transfer(bmRequestType: req_type,
                                     bRequest: COMMAND_SENDSTRING, wValue: 0x0,
                                     wIndex: @accessory.keys.index(k), dataOut: s)

        # TODO: Manage an exception there. This should terminate the program.
        logger.error "Failed to send #{k} string" unless r == s.size
      end
    end
    private :send_accessory_id

    # Sends AOA protocol start command to the device
    # @api private
    # @return [Fixnum]
    def send_start
      logger.debug 'send_start'
      req_type = LIBUSB::ENDPOINT_OUT | LIBUSB::REQUEST_TYPE_VENDOR
      res = @handle.control_transfer(bmRequestType: req_type,
                                     bRequest: COMMAND_START, wValue: 0x0,
                                     wIndex: 0x0, dataOut: nil)
    end
    private :send_start

    # Retrieves an AOAP device by its serial number
    # @api private
    # @param [String] sn
    #   The serial number of the device to be found.
    # @return [LIBUSB::Device] the device matching the given serial number.
    def get_device(sn)
      device = @context.devices(idVendor: GOOGLE_VID).collect do |d|
        d.serial_number == sn ? d : nil
      end.compact.first
    end

    # @api private
    # @return [true, false] true for success, false otherwise.
    def wait_and_retrieve_by_serial(sn)
      sleep Ligo::getDelay
      # The device should now reappear on the usb bus with the Google vendor id.
      # We retrieve it by using its serial number.
      device = get_device(sn)

      if device
        # Retrieve new pointers (check if the old ones should be dereferenced)
        @pDev = device.pDev
        @pDevDesc = device.pDevDesc
        true
      else
        logger.error ['Failed to retrieve the device after switching to ',
                      'accessory mode. This may be due to a lack of proper ',
                      'permissions ⇒ check your udev rules.', "\n",
                      'The Google vendor id rule may look like:', "\n",
                      'SUBSYSTEM=="usb", ATTR{idVendor}=="18d1", ',
                      'MODE="0666", GROUP="plugdev"'
                     ].join
        false
      end
    end
    private :wait_and_retrieve_by_serial

  end

end


lib/ligo/context.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Ligo

  # This class is a derivative work of `LIBUSB::Context` as included in
  #   [LIBUSB](https://github.com/larskanis/libusb), written by Lars Kanis and
  #   released under the LGPLv3.
  # @author Renaud AUBIN
  # @api public
  class Context < LIBUSB::Context
    include LIBUSB

    # @api private
    # Returns the list of AOAP-compatible devices
    # @return [Array<Ligo::Device>] the list of AOAP-compatible devices
    #   currently connected on the USB bus.
    def device_list
      pppDevs = FFI::MemoryPointer.new :pointer
      size = Call.libusb_get_device_list(@ctx, pppDevs)
      ppDevs = pppDevs.read_pointer
      pDevs = []
      size.times do |devi|
        pDev = ppDevs.get_pointer(devi*FFI.type_size(:pointer))
        # Use Ligo::Device instead of LIBUSB::Device
        device = Ligo::Device.new(self, pDev)
        if VENDOR_IDS.include?(device.idVendor)
          begin
            # Include only AOAP compatible devices
            pDevs << device if device.aoap?
          rescue LIBUSB::ERROR_ACCESS
            # @todo Do something about this exception, log at least!
          end
        end
      end
      Call.libusb_free_device_list(ppDevs, 1)
      pDevs
    end
    private :device_list
  end

end


lib/ligo/logging.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'logger'

module Ligo

  # Logging module
  #
  # This module enables to share the same logger between all the Ligo classes.
  module Logging

    def logger
      @logger ||= Logging.logger_for(self.class.name)
    end

    # Use a hash class-ivar to cache a unique Logger per class:
    @loggers = {}
    @out = STDOUT

    class << self
      def logger_for(classname)
        @loggers[classname] ||= configure_logger_for(classname)
      end

      def configure_logger_for(classname)
        logger = Logger.new(@out)
        logger.progname = classname
        logger
      end

      #
      def configure_logger_output(logout)
        @out = logout if logout != 'STDOUT'
      end
    end

  end

end


lib/ligo/accessory.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Ligo

  # A virtual accessory to interact via usb with an android device according
  # to the Android Open Accessory Protocol.
  #
  # @see http://source.android.com/tech/accessories/index.html
  # @see http://source.android.com/tech/accessories/aoap/aoa.html
  # @see
  #  http://developer.android.com/guide/topics/connectivity/usb/accessory.html
  #
  # @author Renaud AUBIN
  # @api public
  class Accessory
    include Logging

    # The default id used to initialize a new accessory if none is provided to
    # the constructor
    DEFAULT_ID = {
      manufacturer: 'ligō',
      model:        'Demo',
      description:  'ligō virtual accessory',
      version:      '1.0',
      uri:          'https://github.com/nibua-r/ligo#readme',
      # 'ligō 1.0'.each_byte {|c| print c.to_i.to_s(16), '' }
      serial:       '6c6967c58d20312e30'
    }

    # Returns the full identifying information hash
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.id
    #   # => {:manufacturer=>"ligō",
    #   #     :model=>"Demo",
    #   #     :description=>"ligō virtual accessory",
    #   #     :version=>"1.0",
    #   #     :uri=>"https://github.com/nibua-r/ligo#readme",
    #   #     :serial=>"6c6967c58d20312e30"}
    # @return [Hash<Symbol, String>] the full identifying information hash.
    attr_reader :id

    # Returns the `manufacturer name` identifying string
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.manufacturer
    #   # => "ligō"
    # @return [String] the `manufacturer name` identifying string.
    attr_reader :manufacturer

    # Returns the `model name` identifying string
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.model
    #   # => "Demo"
    # @return [String] the `model name` identifying string.
    attr_reader :model

    # Returns the `description` identifying string
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.description
    #   # => "ligō virtual accessory"
    # @return [String] the `description` identifying string.
    attr_reader :description

    # Returns the `version` identifying string
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.version
    #   # => "1.0"
    # @return [String] the `version` identifying string.
    attr_reader :version

    # Returns the `uri` identifying string
    # @example
    #   acc = Ligo::Accessory.new
    #   acc.uri
    #   # => "https://github.com/nibua-r/ligo#readme"
    # @return [String] the `uri` identifying string.
    attr_reader :uri

    # Returns the `serial` identifying string
    # @example
    #   accessory = Ligo::Accessory.new
    #   accessory.serial
    #   # => "6c6967c58d20312e30"
    # @return [String] the `serial` identifying string.
    attr_reader :serial

    # Returns a new {Accessory} initialized with the provided identification
    #
    # @param [Hash<Symbol, String>] id
    #   The accessory identifying information as a hash.
    # @raise [ArgumentError] if the provided id is not a Hash or if one of the
    #   identifying string is missing.
    # @example
    #   new_id =
    #   {
    #     manufacturer: 'MyVeryBigCompany Corp.',
    #     model:        'AwesomeProduct',
    #     description:  'Who cares about description! ← 😠',
    #     version:      '0.0',
    #     uri:          'http://www.foo.bar/awesome_product',
    #     serial:       '⚀⚁⚂⚃⚄⚅012345678'
    #   }
    #   accessory = Ligo::Accessory.new(new_id)
    def initialize(id = DEFAULT_ID)

      unless id.is_a? Hash
        raise ArgumentError, '#new must be called with a Hash'
      end

      required_ids = [:manufacturer,
                      :model, :description,
                      :version,
                      :uri,
                      :serial]

      required_ids.each do |sym|
        raise ArgumentError, "Missing argument: #{sym}" unless id.include? sym
      end

      id.each do |k, v|
        raise ArgumentError, "#{k} is not a String" unless v.is_a? String
        raise ArgumentError, "#{k} must not be empty" if v.empty?
        if v.bytesize > 255
          raise ArgumentError, "#{k} must contain at most 255 bytes"
        end
        instance_variable_set "@#{k}", v unless v.nil?
      end

      @id = id
      logger.debug self.inspect
    end

    # Returns {#id}.each
    # @return block execution.
    # @example
    #   accessory.each do |k,v|
    #     puts "#{k} = #{v}"
    #   end
    #   # manufacturer = ligō
    #   # model = Demo
    #   # description = ligō virtual accessory
    #   # version = 1.0
    #   # uri = https://github.com/nibua-r/ligo#readme
    #   # serial = 6c6967c58d20312e30
    def each(&block)
      @id.each(&block)
    end

    # Returns {#id}.keys
    # @return [Array] {#id}.keys.
    # @example
    #   accessory.keys.inspect
    #   # => "[:manufacturer, :model, :description, :version, :uri, :serial]"
    def keys
      @id.keys
    end
  end

end


lib/ligo/constants.rb
# -*- coding: utf-8; fill-column: 80 -*-
#
# Copyright (c) 2012, 2013 Renaud AUBIN
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module Ligo

  # Android devices Vendor Ids
  # @see android source code system/core/adb/usb_vendor.c
  VENDOR_IDS = [
                GOOGLE_VID          = 0x18d1,
                INTEL_VID           = 0x8087,
                HTC_VID             = 0x0bb4,
                SAMSUNG_VID         = 0x04e8,
                MOTOROLA_VID        = 0x22b8,
                LGE_VID             = 0x1004,
                HUAWEI_VID          = 0x12D1,
                ACER_VID            = 0x0502,
                SONY_ERICSSON_VID   = 0x0FCE,
                FOXCONN_VID         = 0x0489,
                DELL_VID            = 0x413c,
                NVIDIA_VID          = 0x0955,
                GARMIN_ASUS_VID     = 0x091E,
                SHARP_VID           = 0x04dd,
                ZTE_VID             = 0x19D2,
                KYOCERA_VID         = 0x0482,
                PANTECH_VID         = 0x10A9,
                QUALCOMM_VID        = 0x05c6,
                OTGV_VID            = 0x2257,
                NEC_VID             = 0x0409,
                PMC_VID             = 0x04DA,
                TOSHIBA_VID         = 0x0930,
                SK_TELESYS_VID      = 0x1F53,
                KT_TECH_VID         = 0x2116,
                ASUS_VID            = 0x0b05,
                PHILIPS_VID         = 0x0471,
                TI_VID              = 0x0451,
                FUNAI_VID           = 0x0F1C,
                GIGABYTE_VID        = 0x0414,
                IRIVER_VID          = 0x2420,
                COMPAL_VID          = 0x1219,
                T_AND_A_VID         = 0x1BBB,
                LENOVOMOBILE_VID    = 0x2006,
                LENOVO_VID          = 0x17EF,
                VIZIO_VID           = 0xE040,
                K_TOUCH_VID         = 0x24E3,
                PEGATRON_VID        = 0x1D4D,
                ARCHOS_VID          = 0x0E79,
                POSITIVO_VID        = 0x1662,
                FUJITSU_VID         = 0x04C5,
                LUMIGON_VID         = 0x25E3,
                QUANTA_VID          = 0x0408,
                INQ_MOBILE_VID      = 0x2314,
                SONY_VID            = 0x054C,
                YULONG_COOLPAD_VID  = 0x1EBF
               ]

  # Accessory Protocol allowed product ids
  GOOGLE_PIDS = *(0x2d00..0x2d05)

  ## Accessory Protocol Commands
  # Accessory Protocol "Get Protocol" command
  COMMAND_GETPROTOCOL = 51
  # Accessory Protocol "Send Id String" command
  COMMAND_SENDSTRING  = 52
  # Accessory Protocol "Accessory Mode Start Up" commands
  COMMAND_START       = 53

end
